File: src_bak/tizenclaw/infra/dbus_probe.cc
/*
 * Copyright (c) 2026 Samsung Electronics Co., Ltd All Rights Reserved
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "dbus_probe.hh"

#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

#include <atomic>

#include "../../common/logging.hh"

namespace {

enum class ProbeState {
  kUntested = 0,
  kAvailable = 1,
  kUnavailable = 2,
};

std::atomic<ProbeState> g_state{ProbeState::kUntested};

// The child process attempts to open the D-Bus
// system bus socket.  If libdbuspolicy1 triggers
// assert(false) inside udesc::init_once(), the
// child will be killed by SIGABRT.
// We intentionally avoid linking libdbus here;
// instead we check connectivity by trying to open
// the well-known system bus socket path.
bool RunProbe() {
  pid_t pid = fork();
  if (pid < 0) {
    // fork failed — assume D-Bus is available
    // so we don't silently disable features.
    return true;
  }

  if (pid == 0) {
    // Child: attempt to connect to the D-Bus
    // system bus socket.  The glib/dbus libraries
    // loaded via Tizen APIs will trigger
    // libdbuspolicy1 init on first use.
    // We use a lightweight approach: just check
    // the existence and accessibility of the
    // D-Bus system bus socket.
    const char* bus_addr =
        "/run/dbus/system_bus_socket";
    if (access(bus_addr, R_OK | W_OK) == 0) {
      _exit(0);  // socket accessible
    }
    _exit(1);  // socket not accessible
  }

  // Parent: wait for child result.
  int status = 0;
  if (waitpid(pid, &status, 0) < 0) {
    return true;  // waitpid error — assume ok
  }

  if (WIFSIGNALED(status)) {
    int sig = WTERMSIG(status);
    LOG(WARNING) << "DbusProbe: child killed by "
                 << "signal " << sig
                 << (sig == 6 ? " (SIGABRT)" : "");
    return false;
  }

  if (WIFEXITED(status)) {
    int code = WEXITSTATUS(status);
    if (code != 0) {
      LOG(WARNING) << "DbusProbe: D-Bus socket "
                   << "not accessible (exit="
                   << code << ")";
      return false;
    }
    return true;
  }

  return true;
}

}  // namespace

namespace tizenclaw {

bool DbusProbe::IsAvailable() {
  ProbeState expected = ProbeState::kUntested;
  if (g_state.compare_exchange_strong(
          expected, ProbeState::kAvailable)) {
    // First caller — run the probe.
    bool ok = RunProbe();
    g_state.store(ok ? ProbeState::kAvailable
                     : ProbeState::kUnavailable);
    LOG(INFO) << "DbusProbe: D-Bus system bus "
              << (ok ? "available" : "unavailable");
    return ok;
  }

  // Wait for probe to finish if another thread
  // is running it concurrently (rare).
  while (g_state.load() == ProbeState::kUntested) {
    usleep(1000);
  }

  return g_state.load() == ProbeState::kAvailable;
}

}  // namespace tizenclaw

File: src/tizenclaw/src/generic/infra/dbus_probe.rs
//! D-Bus probe — checks whether the D-Bus system bus is accessible.
//!
//! Ported from C++ DbusProbe: uses fork() to safely probe the socket,
//! guarding against libdbuspolicy1 SIGABRT crashes.

use std::sync::atomic::{AtomicU8, Ordering};

const STATE_UNTESTED: u8 = 0;
const STATE_AVAILABLE: u8 = 1;
const STATE_UNAVAILABLE: u8 = 2;

static PROBE_STATE: AtomicU8 = AtomicU8::new(STATE_UNTESTED);

const DBUS_SYSTEM_BUS_SOCKET: &str = "/run/dbus/system_bus_socket";

/// Run the probe in a child process to safely test D-Bus availability.
fn run_probe() -> bool {
    unsafe {
        let pid = libc::fork();
        if pid < 0 {
            // fork failed — assume available
            return true;
        }

        if pid == 0 {
            // Child: check socket accessibility
            let path = std::ffi::CString::new(DBUS_SYSTEM_BUS_SOCKET).unwrap();
            if libc::access(path.as_ptr(), libc::R_OK | libc::W_OK) == 0 {
                libc::_exit(0);
            }
            libc::_exit(1);
        }

        // Parent: wait for child
        let mut status: libc::c_int = 0;
        if libc::waitpid(pid, &mut status, 0) < 0 {
            return true; // waitpid error — assume ok
        }

        // Check if child was killed by signal (e.g., SIGABRT)
        if libc::WIFSIGNALED(status) {
            let sig = libc::WTERMSIG(status);
            log::warn!(
                "DbusProbe: child killed by signal {}{}",
                sig,
                if sig == 6 { " (SIGABRT)" } else { "" }
            );
            return false;
        }

        if libc::WIFEXITED(status) {
            let code = libc::WEXITSTATUS(status);
            if code != 0 {
                log::warn!("DbusProbe: D-Bus socket not accessible (exit={})", code);
                return false;
            }
            return true;
        }

        true
    }
}

/// Check if D-Bus system bus is available. Result is cached.
pub fn is_available() -> bool {
    let prev = PROBE_STATE.load(Ordering::SeqCst);
    if prev != STATE_UNTESTED {
        return prev == STATE_AVAILABLE;
    }

    // First caller runs the probe
    if PROBE_STATE.compare_exchange(STATE_UNTESTED, STATE_AVAILABLE, Ordering::SeqCst, Ordering::SeqCst).is_ok() {
        let ok = run_probe();
        PROBE_STATE.store(
            if ok { STATE_AVAILABLE } else { STATE_UNAVAILABLE },
            Ordering::SeqCst,
        );
        log::debug!(
            "DbusProbe: D-Bus system bus {}",
            if ok { "available" } else { "unavailable" }
        );
        return ok;
    }

    // Another thread is probing — wait
    while PROBE_STATE.load(Ordering::SeqCst) == STATE_UNTESTED {
        std::thread::sleep(std::time::Duration::from_millis(1));
    }

    PROBE_STATE.load(Ordering::SeqCst) == STATE_AVAILABLE
}
